resources/bundles/org.eclipse.core.resources/src/org/eclipse/core/internal/localstore/FileSystemResourceManager.java

```diff
@@ -1099,15 +1099,16 @@ public void updateLocalSync(ResourceInfo info, long localSyncInfo) {
 	}
 
 	/**
-	 * The target must exist in the workspace. The content InputStream is
-	 * closed even if the method fails. If the force flag is false we only write
-	 * the file if it does not exist or if it is already local and the timestamp
-	 * has NOT changed since last synchronization, otherwise a CoreException
-	 * is thrown.
+	 * The target must exist in the workspace and must remain existing while
+	 * executing this method. The content InputStream is closed even if the method
+	 * fails. If the force flag is false we only write the file if it does not exist
+	 * or if it is already local and the timestamp has NOT changed since last
+	 * synchronization, otherwise a CoreException is thrown.
 	 */
 	public void write(IFile target, InputStream content, IFileInfo fileInfo, int updateFlags, boolean append, IProgressMonitor monitor) throws CoreException {
 		SubMonitor subMonitor = SubMonitor.convert(monitor, 4);
 		try {
+			assertExists(target);
```

Comment:
Did you consider throwing a CoreException to maintain the contract of the API, instead of an IllegalStateException?

Reply:
@jukzi Maybe this question is rather directed to you, since you worked on https://git.eclipse.org/r/c/platform/eclipse.platform.resources/+/176136

Reply:
Good point. I have to admit that I am not sure whether I understood the contract correctly. I expected the existance of `IFile target` to be a precondition for executing the method, which is why an `IllegalStateException` would make sense to me if called with a non-existing `target`. I expected the condition regarding throwing a `CoreException` to refer to the static information in the given `fileInfo` or in the `target`, given that the latter is present.

But you are right that in certain situations the method will throw a different exception now, so I am completely open to throwing a `CoreException` instead. However, there is another test case that expects an `IllegalStateException` to be thrown when passing a non-existing target (see https://github.com/eclipse-platform/eclipse.platform/issues/103#issuecomment-1285582896).

Either way, as soon as we have agreed we may also improve the contract to clarify its meaning.

Reply:
@jukzi would be great to have your feedback on which kind of exception you would expect to be thrown here (in case you have an opinion on that). thanks!

Reply:
> Changing the contract is bad even if it is an improvement in some cases - there may be code that relies on the contract.

True. What I actually meant was that we should adapt the Javadoc so that it properly represents the existing behavior of the method, since the specification is currently not very precise.

> The javadoc clearly specifies CoreException should be thrown if file is missing.

That is one possible interpretation, but I would say that it "clearly specifies" this. Throwing a `CoreException` is specified for the case that the force flag is false and the other mentioned conditions are not fulfilled. The Javadoc does not make a statement about the exception thrown in the case that the target is not existing. Actually, it defines this as a precondition, which is why I find an `IllegalStateException` appropriate here. In addition, the current code throws an `IllegalStateException` at several points where the resource info is null.

> However as far i remember the root for the IllegalStateException was that platform was not really sure if the file exists or not. A proper fix would be to avoid to get into such state - just don't know how exactly that is reached. 

The reason I saw so far is that the file is concurrently removed and a concurrent workspace refresh is performed, such that the file is not represented in the workspace tree anymore. So maybe a proper/better solution would be to acquire a workspace lock during the write operation or to even wrap it into a workspace operation. But I fear that will at least introduce a performance issue and may, in the worst case, even lead to deadlocks.